[그리디] 김지우 Spring MVC 3,4단계 제출합니다

Ji-Woo-Kim opened the pull request:
동훈님, 지난 미션에 이어 이번 주도 잘 부탁드립니다~! 🙇‍♀️ 

---

**지난 주와 달라진 점, 미션 진행하면서 신경 쓴 부분**들에 대해 간략히 정리해보자면 다음과 같습니다 :
### ✅ controller 분리
 `PageController`와 `ReservationController`로 분리했습니다.
 `PageController`에선 단순히 페이지의 이동만을 하는 요청들만 모아 두고, 
  예약 목록에 관한 여러 기능(조회, 추가, 삭제)에 대한 요청은 `ReservationController`에 두는게 관심사를 나눠서 설계하는 이점이 있을 것 같아 분리해보았습니다. 
  이렇게 분리한 것에 대한 동훈님의 의견이 궁금합니다! 
  (잘 분리한 것 같은지, 아님 단순히 목록 페이지로 넘어간다 해도 reservationController에 두는게 더 괜찮은지..)

### ✅ dto 구현
 요청/응답에 대한 dto를 도입해봤습니다.
 아래 설명할 예외처리와도 이어지는 부분인데요, `ReservationRequestDto`에 Bean Validation 어노테이션(@NotNull, @NotBlank)을 추가하여 유효성 검사를 수행하도록 설정하였습니다.

### ✅ 예외 처리
 예외처리 로직을 짜보는게 이번이 처음이어서 초록스터디 학습테스트를 기반으로 작성해보았는데요, 
 놓친 부분이나 좀 더 생각해서 다듬을 부분이 있다면 공유해주시면 감사하겠습니다! 

---

### **해결이 안 된 부분 🥲** :
 nextstep에서 제공해준 테스트코드 3-4단계가 tests failed이 뜨는데 해결이 안 됐습니다..ㅜ
특히 3단계 테스트 실의 원인을 보니, Controller에서 임의로 생성된 3개의 데이터 때문에 POST → DELETE 후 예상하는 데이터 개수와 실제 남아있는 데이터 개수가 일치하지 않는 것 같습니다. 이를 해결하기 위해 테스트 전 초기화 로직을 추가했으나, 여전히 해결되지 않았습니다.
혹시 controller 로직 자체에 문제가 있는 걸까요..? (실제 localhost:8080에서 실행할 때는 정상적으로 동작하는 것 같습니다만...
테스트에서만 실패하는 이유가 궁금합니다.)

---

제가 언급한 부분 이외에 개선방안이나, 문제점, 추가로 알아보고 학습해야 할 것들이 있다면 편하게 남겨주시길 바랍니다!

감사합니다 😃

Ji-Woo-Kim committed:
feat: 1단계 구현
- Gradle 의존성 추가
- Controller 생성(HomeController)

Ji-Woo-Kim committed:
refactor: controller 패키지 추가

Ji-Woo-Kim committed:
feat: 2단계 구현
- Reservation 도메인 생성
- ReservationController 생성
- 예약 관리 페이지, 목록 조회 구현

Ji-Woo-Kim committed:
Create README.md

Ji-Woo-Kim committed:
feat: 3단계 구현
- Controller 분리 : PageController와 ReservationController
- 예약 추가/취소 기능 구현

Ji-Woo-Kim committed:
feat: 4단계
- 예외처리 로직 추가
- Dto 구현

Ji-Woo-Kim committed:
refactor: DTO validation 추가 및 예외 처리 보완
- ReservationRequestDto에 @NotNull, @NotBlank, @Valid 어노테이션 추가
- GlobalExceptionHandler에 MethodArgumentNotValidException 처리 로직 추가
- build.gradle validation 의존성 추가

Ji-Woo-Kim committed:
Merge branch 'ji-woo-kim' into step3-4

Ji-Woo-Kim committed:
Update README.md

shin-mallang commented on src/main/java/roomescape/controller/PageController.java:
```diff
@@ -0,0 +1,18 @@
+        return "home";
+    }
+
+    @GetMapping("/reservation")
```
질문 남겨주신 부분인데, 현재 상황에서는 PageController 에서 reservation 페이지를 보여주는 역할을 수행해도 문제없을 거 같습니다
다만 이후 보여주어야 할 페이지가 많다면 PageController 에 너무 많은 메서드가 들어갈 거 같아 그런 경우라면 ReservationController 에 옮겨주어도 좋을 거 같아요.
저는 개인적으로 ReservationController 에 둘 거 같은데요, 이유는 예약 관련 화면을 보여주는 기능을 찾을 때, PageController 보다는 ReservationController 를 먼저 확인하게 될 거 같기 때문입니닷!

Ji-Woo-Kim replied:
동훈님 의견을 듣고 나니 ,
`Controller`를 api domain에 따라 네이밍하고 그에 맞는 관련 로직을 담아두는 것이 더 유용한 설계라고 생각되어 수정했습니다!
- `PageController` -> `HomeController` (홈페이지만 랜더링)
- `GET ("/reservation")메서드` 
   `ReservationController`으로 이동

shin-mallang commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -1,30 +1,65 @@
-        return reservations;
+    @PostMapping
+    public ResponseEntity<ReservationResponseDto> create(@Valid @RequestBody ReservationRequestDto requestDto) {
+        if (requestDto.getName().isEmpty() || requestDto.getDate() == null || requestDto.getTime() == null) {
```
@Valid 를 통해서 이러한 상황에 대한 예외처리를 진행해주신 것 같은데, 코드에 추가해주신 이유가 있으실까요?

Ji-Woo-Kim replied:
4단계 진행하면서 처음엔 하나하나 조건문으로 예외처리 하다가 이후 @Valid에 대해 알게 되었어요. 어노테이션 적용 후에 메서드 내 유효성 검증 로직을 삭제했어야 하는데 그냥 넘어가 버렸네요... ㅜㅜ

shin-mallang replied:
아하 그런 사정이 있었군요 ㅋㅋㅋㅋ

shin-mallang commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,42 @@
+        return this.time;
+    }
+
+    public Reservation toEntity(Long id) {
```
Reservation.fromDto(dto) 등과 같은 식으로 구현할수도 있었을 거 같은데,
이러한 방식으로 구현하신 이유가 궁금합니다~

Ji-Woo-Kim replied:
정팩메를 활용하라는 의견이시네요!

일단 이번에 그리디 들어와서 자바 미션을 진행하면서 정팩메에 대해 처음 알게 됐습니다. 그래서 아직 어느 상황에 활용해야 하는지 등의 판단이 능숙하지 않아 미처 정팩메를 사용해야 겠다는 생각은 하지 못 했어요. 🥹 
그리고 객체로 변환해야 겠다는 생각에 빠져 DTO에서 바로 Entity로 변환하도록 toEntity() 메서드를 사용했는데요.

동훈님 의견을 받고 두 방식을 비교해보니,
`toEntity()`를 사용하면 DTO가 Entity 생성을 책임지게 되어 너무 강한 결합이 생성된다는 문제가 있다고 판단되었습니다.
반면, 정팩메를 사용하면 객체 본인이 생성 역할을 갖기 때문에 책임 분리가 더 명확해지는 설계가 될 수 있을 것으로 보이네요!

정팩메를 이용하도록 수정하겠습니다 :)

shin-mallang replied:
> 정팩메를 활용하라는 의견이시네요!

헉, 그렇게 하지 않았던 것에 대한 단순 질문이었습니다!

그럼에도 불구하고 바꿔주셨기 때문에 이야기를 좀 더 진행해보면 좋을 거 같은데요,

> toEntity()를 사용하면 DTO가 Entity 생성을 책임지게 되어 너무 강한 결합이 생성된다는 문제가 있다고 판단되었습니다.
반면, 정팩메를 사용하면 객체 본인이 생성 역할을 갖기 때문에 책임 분리가 더 명확해지는 설계가 될 수 있을 것으로 보이네요!

Dto가 Entity와의 강한 결합을 가지면 안 좋다고 생각한 이유는 무엇인지 궁금하고,
또한 설명해주신 것에 따르면 Entity.from(dto); 와 같은 구조로 설계를 변경한 경우, Entity에서 Dto에 강한 결합성을 가지게 되는 것 같은데, 요렇게 진행했을 때의 문제점은 따로 없을까요?

Ji-Woo-Kim replied:
우선 처음 남겨주신 리뷰를 코드에 반영한 이유는
말씀해주신 것처럼 'DTO가 Entity 생성을 책임지게 되어 너무 강한 결합이 생성된다는 문제'를 해결하고 싶다는 생각이 컸습니다.

`DTO`는 각 계층에 데이터를 전달하기 위한 것으로 목적이 **데이터 전달**이라고 생각을 하는데요, 그래서 이 목적에 맞게 dto 내부엔 최대한 도메인 로직이 안 담기도록 하는 것이 좋다고 생각합니다.

그래서 리뷰를 보고 난 뒤 `entity` 본인이 스스로의 생성 로직을 담도록 정팩메를 이용하는 것이 더 괜찮은 설계라고 생각했는데요, 이 방식 또한 entity와 dto 간의 강한 결합성을 해소시켜주진 않는 것을 인지했습니다. 
또 entity가 dto에 의존함에 따라 dto의 변화에 맞춰 entity도 바꿔줘야 하는 등 유지복수 측면의 문제도 있다고 생각합니다.
특히, 이 문제는 dto가 늘어날수록 상당한 문제가 될 것 같습니다.

그러나, 현재는 dto가 한 개밖에 없기 때문에 현재 설계(정팩메)를 유지하는 것도 나쁘지 않다고 생각합니다.
그리고 추후 dto가 늘어난다면 Mapper 클래스를 추가하여 dto-entity 간 결합도를 낮추는 방식을 도입하는 것도 고려해보려 합니다!

혹시 이와 관련해 더 나은 구조나 경험 있으시다면 공유해주세요..!

shin-mallang replied:
> entity가 dto에 의존함에 따라 dto의 변화에 맞춰 entity도 바꿔줘야 하는 등 유지복수 측면의 문제도 있다고 생각합니다.

지우님이 언급해주신 해당 문제 때문에, 저는 개인적으로 dto.toEntity() 방식의 코드를 선호합니다!
dto의 구조가 변경되었을 때 entity가 영향을 받는 것이 그닥 좋은 구조라 생각되지 않아서요!

위 두 방식은 곃국 dto가 entity를 의존하는 것과, entity가 dto를 의존하는 것 중 하나를 고르는 것과 같은 문제라 생각되는데요,

dto의 특성 상 어쩔 수 없이 dto -> entity 로의 논리적 의존성이 존재한다고 생각합니다.

결론적으로 저는 dto.entity의 방식을 선호하기는 합니닷!

Ji-Woo-Kim replied:
관련된 내용으로 좀 더 서치해보니 
규모가 크지 않은 플젝에서는 `dto.toEntity()` 방식을 많이 사용하시는 느낌을 받았습니다.

그리고 제가 리팩하여 적용한 `entity.fromDto()` 방식은 결합도를 낮추기 위해 Mapper 클래스를 거의 필수적(?)으로 사용하는 것 같아요.

현재 플젝은 아직 규모가 크지 않아서 두 방식 중 어느 하나를 명확히 사용하는 것이 좋겠다는 판단을 내리기가 애매하네요 ..😅 

일단 코드 리팩은 따로 하지 않겠습니다.! 
이 리뷰를 인지하면서 추후 미션들을 하다보면 _dto가 entity를 의존하는 것과, entity가 dto를 의존하는 것_에 대한 판단 기준이 세워질 것 같은 생각입니다.

shin-mallang commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -1,30 +1,65 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
 
+@RequestMapping(value = "/reservations", produces = "application/json; charset=UTF-8")
```
, produces = "application/json; charset=UTF-8" 옵션이 무엇이고 꼭 필요한지? 없으면 어떻게 되는지 궁금합니다!

Ji-Woo-Kim replied:
우선 produces 옵션은 서버가 응답할 타입을 직접 지정해줘서 
응답 헤더에 이 타입이 설정이 되도록 하는 것으로 알고 잇습니다..!

테스트 코드의 오류가 뭔지 스스로 찾지 못 해서 결국 오류 로그를 gpt에 넘겨서 확인한 뒤 추가하게 된 코드인데요.
결론적으로 Spring Boot 2.0 이상에서는 application/json을 보낼 때 기본적으로 UTF-8이 설정(텍스트 인코딩할 때 한글이 깨지지 않도록 해줍니다!)되기에 필요없는 코드라고 판단했었습니다.

그래서 제거한 뒤 커밋하려고 했었는데 그대로 커밋을 해버렸네요.. 😓

shin-mallang commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Map;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
요게 처음 작성해보신 예외처리 로직일까욤??
너무 잘 작성해 주셨네여!!

shin-mallang commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@ControllerAdvice
```
@ControllerAdvice 가 어떤 역할을 하는 어노테이션인가요?

Ji-Woo-Kim replied:
@ControllerAdvice는 
- @ExceptionHandler가 붙은 예외 처리 메서드의 적용 범위를 controller 전역으로 넓혀줍니다.
- 여러 속성 설정으로 개발자가 원하는 곳에만 적용되도록 수정할 수도 있습니다.
- 추가로, @ResponseBody와 결합한 형태인 @RestControllerAdvice는 Api 응답 시 json 형식으로 응답합니다. 

아래 링크 참고하여 정리해보았습니다~!
https://tecoble.techcourse.co.kr/post/2023-05-03-ExceptionHandler-ControllerAdvice/

shin-mallang commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,33 @@
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(InvalidReservationRequestException.class)
```
이런 구조라면 예외를 추가할 때마다 GlobalExceptionHandler 에 추가되는 구조일 거 같아요~
예외가 새롭게 추가되더라도, GlobalExceptionHandler 에서는 코드의 변경 없이 일관적으로 예외처리가 진행되게끔 하는 방법은 없을까요?

Ji-Woo-Kim replied:
이 질문에 대해 아직 제 스스로 방법을 떠올리기 어려워서 스프링 예외 처리를 서치하면서 동훈님 질문의 의도가 뭔지 파악해보려고 해봤는데요..!

`ResponseEntityExceptionHandler`를 상속받아 Spring이 제공하는 예외와 매핑하는 방식을 이용하라는 말씀이실까요?
그치만 이 방식대로 한다면 에러 메세지를 또 오버라이드 해서 설정해주어야 한다고 해서 코드를 변경하는 번거로움이 해결되지 않는 것 같다고 느껴집니다.

아니면 다른 방법으로는, 각 예외들의 상위 클래스를 구현해서 마찬가지로 상속을 이용하여 처리하는 방법도 발견했습니다.

혹시 다른 걸 의도하셨다면 힌트가 될 수 있는 키워드가 있을까요?! 🥹

shin-mallang replied:
> 아니면 다른 방법으로는, 각 예외들의 상위 클래스를 구현해서 마찬가지로 상속을 이용하여 처리하는 방법도 발견했습니다.

요걸 의도하긴 했습니다~!~! 👍👍

Ji-Woo-Kim replied:
각 예외들의 상위 클래스를 구현해서 상속을 이용하여 처리하는 방법을 도입한다면 예외들을 그룹화할 때의 기준을 어떤 식으로 잡는게 좋을까요? 🤔 

- 도메인 기준으로...
- 패키지(계층) 기준으로...
- HTTP 상태코드 기준으로...
- 에러의 원인(유효성, 인증, 권한 등)을 기준으로...
- (다른 기준..?)

제가 고민해보았을 땐 도메인 기준으로 나눴을 때 어떤 상황에서 발생한 예외임을 뚜렷하게 전달해주는 것 같아서 나쁘지 않은 방법 같다고 느껴집니다. 
그치만, 다른 도메인이라 하더라도 발생하는 에러 유형이 동일한 경우도 생각보다 빈번하기 때문에 코드의 중복성이 높아지는 문제점이 있어 보이는데요..!

이 부분이 개발자 판단에 따라 설계가 다양해지는지 일반적으로 분리하는 기준이 있는지 모르겠어서 질문 남겨드립니다 😅

shin-mallang replied:
> 각 예외들의 상위 클래스를 구현해서 상속을 이용하여 처리하는 방법을 도입한다면 예외들을 그룹화할 때의 기준을 어떤 식으로 잡는게 좋을까요? 🤔
도메인 기준으로...
패키지(계층) 기준으로...
HTTP 상태코드 기준으로...
에러의 원인(유효성, 인증, 권한 등)을 기준으로...
(다른 기준..?)

저는 개인적으로 도메인 기준도 괜찮고, HTTP 상태코드 기준도 괜찮은 거 같습니닷~

> 그치만, 다른 도메인이라 하더라도 발생하는 에러 유형이 동일한 경우도 생각보다 빈번하기 때문에 코드의 중복성이 높아지는 문제점이 있어 보이는데요..!

요런 경우에는 예외에 에러 유형 필드를 두고, 해당 예외가 어떤 유형의 예외인지를 명시해주는 방법을 통해 중복을 일정 수준 줄일 수 있다고 생각합니닷!

다만 아직 저도 뭐가 정답이고, 제일 좋다에 대해서는 확실치 않아서 말씀을 못 드리겠네요..!

프로젝트 할 때마다 매번 예외 처리 방식을 바꿔가면서 테스트해보는 것 같습니닷!!

지우님도 한번 두 방법 다 해 보시고 더 편한 방법을 선택하시는 것도..? 좋아보여요?

혹은 요런건 다른 사람들과 의견 공유를 추가로 진행해봐도 좋을 거 같습니닷 :)

Ji-Woo-Kim replied:
커스텀 예외들의 상위 클래스를 도메인 기준으로 잡아 만들었습니다.
: `ReservationException`
여기에 HTTP 상태코드와 에러타입을 필드로 두는 방식을 선택했습니다.

그리고 기존 커스텀 예외들을 상속구조로 변경한 뒤, GlobalExceptionHandler에서 중복되는 구조를 하나의 메서드로 축소 시켰습니다!

확실히 더 깔끔해진 느낌이 들어요 👍 

---

우선 이 방식이 좋은 선택인 것 같아서 적용해보았습니다!
말씀주신 것처럼 다른 분들의 의견을 구해보거나 다른 pr들도 관찰해봐야 겠어요 🤓

shin-mallang reviewed (commented):
안녕하세요 지우님!
리뷰가 너무 늦었네요. 죄송합니다 ㅠㅠ.

코드 너무 잘 작성해 주셔서 리뷰하기 너무 편했습니다~
간단히 의견 남겼으니 답변 부탁드려요!

추가로 테스트코드 관련해서는 제가 지금 노트북 사용이 어려운 상황이라 내일 중으로 확인 후 답글 달아두도록 하겠습니다!

너무 고생하셨어요!

shin-mallang commented:
테스트 관련해서는 3단계의 경우 ReservationController 에서 초기에 넣어주는 3개의 Reservation 들을 제거해주면 해결될 거 같아요~
(그렇게 되면 테스트 2단계는 실패하게 되는데, 요건 2단계의 검증부를 고쳐서 해결해주셔도 되고 그대로 남겨주셔도 괜찮을 거 같습니다!)

그리고 고치더라도 4단계 자체는 통과되지 않는데, 원인은 지우님 스스로 잘 파악하실 수 있을 거 같아서 우선 따로 언급하지 않고 넘어가겠습니다!

Ji-Woo-Kim committed:
refactor: Entity, DTO 리뷰 반영
- build.gradle lombok 의존성 추가
- Reservation: Getter, 생성자 어노테이션 추가
- Reservation: 정팩메 적용
- ReservationRequestDto: Data 어노테이션 적용, 메서드 삭제

Ji-Woo-Kim committed:
refactor: 페이지 랜더링용 -> 기본 홈 페이지 컨트롤러

Ji-Woo-Kim committed:
refactor: 리뷰 반영
- 예약 페이지 랜더링 메서드(GET)
- 중복 검증 로직 삭제
- 테스트1-2단계용 임시 데이터셋 제거

Ji-Woo-Kim committed:
refactor: NotFoundReservation 반환 Http 상태코드 404 -> 400으로 변경

Ji-Woo-Kim committed:
test: 불필요한 import 제거
- 삼단계(), 사단계() passed

Ji-Woo-Kim commented:
동훈님, 주말 잘 보내셨나요 🥹 

개인적인 과제들이 계속해서 쌓인 채로 해결이 잘 안 되는 상황이라 예상보다 리뷰 반영이 늦어져서 죄송한 마음입니다..ㅠㅜ
(수요일에 JDBC 미션 PR을 날리기 전에 MVC를 머지하는 것을 목표로 이번 미션 마무리하는 것을 우선적으로 해보려고 합니다..!)

---

리뷰 남겨주신 걸 기반으로 차근차근 리팩토링 해보니
 해결이 안 되던 부분들은 다 해결이 되었습니다! 
막상 해결해보니, 제 생각보다 사소한 부분들이어서 조금 허무한 부분도 있었지만 그래도 해결이 돼서 홀가분 합니다 😄
 
### 관련된 부분은... :
- 삼단계, 사단계 테스트 통과를 위해 controller에 작성되어 있던 임시 데이터들을 삭제했습니다.
  - 이미 1-2단계 미션하면서 pass인 것을 확인했으나, 그 이후로 리팩토링된 부분들이 있어 삭제하기 직전에 다시 실행해보았습니다. 일단계(), 이단계() pass 확인 후에 임시 데이터 코드 삭제했습니다.
- `404(NOT_FOUND)` 반환하던 것을 `400(BAD_REQUEST)`로 수정하여 사단계() pass 확인했습니다
  - 예외 메서드명에 따라 자연스레 NOT_FOUND로 반환하도록 설계했었는데 지난 주에 이 부분을 아예..못 봤네요 (저도 잘 이해가 안 갑니다..)

이외에 확인해주실 부분은 예외 처리 부분에 남겨주신 코멘트에 대한 답변 정도일 것 같습니다!

혹시 추가로 이야기 나눠볼 곳이 있다면 알려주세요~ 😄 
바쁘신 와중에도 리뷰해 주셔서 감사합니다 🙇‍♀️

Ji-Woo-Kim committed:
feat: 커스텀 예외 상위 클래스

Ji-Woo-Kim committed:
refactor: 커스텀 예외 상위 클래스 도입
- 기존 커스텀 예외 상속으로 수정
- GlobalExceptionHandler 중복 구조 축소

Ji-Woo-Kim committed:
refactor: @Getter 추가

Ji-Woo-Kim committed:
feat: 순수 JDBC로 CRUD 구현1
- Create, Read

shin-mallang commented on src/main/java/roomescape/exception/NotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+import org.springframework.http.HttpStatus;
+
+public class NotFoundException extends ReservationException {
```
특정 도메인 예외임을 좀 더 명시적으로 드러내기 위해 NotFoundReservationException 과 같은 네이밍을 사용해도 좋을 거 같아요~

도메인이 많아지면 같은 이름의 예외 클래스가 많아질거 같아서요!

Ji-Woo-Kim replied:
아!! 넵 반영하겠습니다!

Ji-Woo-Kim committed:
refactor: reservation예외 메서드명 변경
